Sources/_OpenAPIGeneratorCore/Hooks/FilteredDocument.swift

```
@@ -0,0 +1,485 @@
+                    try addComponentsReferencedBy(try document.components.lookup(reference))
+                case .b(let schema):
+                    try addComponentsReferencedBy(schema)
+                case .none:
+                    continue
+                }
+            }
+        }
+    }
+
+    mutating func addComponentsReferencedBy(_ response: OpenAPI.Response) throws {
+        if let headers = response.headers {
+            for header in headers.values {
+                try requireHeader(header)
+            }
+        }
+        for content in response.content.values {
+            try addComponentsReferencedBy(content)
+        }
+        for link in response.links.values {
+            try requireLink(link)
+        }
+    }
+
+    mutating func addComponentsReferencedBy(_ content: OpenAPI.Content) throws {
+        if let schema = content.schema {
+            try requireSchema(schema)
+        }
+        if let encoding = content.encoding {
+            for encoding in encoding.values {
+                if let headers = encoding.headers {
+                    for header in headers.values {
+                        try requireHeader(header)
+                    }
+                }
+            }
+        }
+        if let examples = content.examples {
+            for example in examples.values {
+                try requireExample(example)
+            }
+        }
+    }
+
+    mutating func addComponentsReferencedBy(_ content: OpenAPI.Link) throws {}
+
+    mutating func addComponentsReferencedBy(_ content: OpenAPI.Example) throws {}
+}
+
+fileprivate extension OpenAPI.Reference {
+    var internalComponentKey: OpenAPI.ComponentKey {
+        get throws {
+            guard case .internal(.component(name: let name)) = jsonReference else {
+                throw FilteredDocumentBuilderError.cannotResolveInternalReference(absoluteString)
+            }
+            return OpenAPI.ComponentKey(rawValue: name)!
```

Can we throw an error if it fails to init? Looking at the impl, there might be situations where the user's document could cause this to crash, which becomes difficult to debug.